reflect/src/main/java/dagger/reflect/UnlinkedJustInTimeBinding.java

```diff
@@ -32,7 +43,65 @@
     return new LinkedJustInTimeBinding<>(constructor, bindings, membersInjector);
   }
 
+  private Type getTypeKeyForParameter(Type parameterType) {
+    if (parameterType instanceof ParameterizedType) {
+      return findKeyForParameterizedType((ParameterizedType) parameterType);
+    } else {
+      return parameterType;
+    }
+  }
+
+  private TypeUtil.ParameterizedTypeImpl findKeyForParameterizedType(
+      ParameterizedType parameterType) {
+    Type matchingType = matchingParameterizedType(parameterType);
+    return new TypeUtil.ParameterizedTypeImpl(null, parameterType.getRawType(), matchingType);
+  }
+
+  /**
+   * Given a parameterized type `Provider<T>` where `T` is our type parameter placeholder: We first
+   * look at this class's parameterized type declarations to see if we can find a matching `T`. When
+   * we find a matching `T`, we use `T`'s index to find the corresponding concrete type argument in
+   * this binding's concreteTypeArguments.
+   *
+   * @param parameterizedType The parameterized type placeholder to lookup.
+   * @return The matching concrete type for the placeholder.
+   */
+  private Type matchingParameterizedType(ParameterizedType parameterizedType) {
+    Type[] actualTypeArguments = parameterizedType.getActualTypeArguments();
+    if (actualTypeArguments.length > 1) {
+      throw new IllegalStateException(
+          "Multiple parameterized type arguments are not supported yet.");
+    }
+    Type typeToLookup = actualTypeArguments[0];
+    // If this isn't a TypeVariableImpl like <T> than we can just return the parameterized type.
+    if (!(typeToLookup instanceof TypeVariable)) {
+      return parameterizedType;
+    }
+    if (concreteTypeArguments == null) {
+      throw new IllegalStateException(
+          "No concrete type arguments for " + cls + " but needed for " + parameterizedType);
+    }
+    // Iterate through parameterized types declared in this class to find the matching index and
+    // return the corresponding concrete type.
+    for (int i = 0; i < cls.getTypeParameters().length; i++) {
+      if (cls.getTypeParameters()[i] == typeToLookup) {
```

Comment:
Is reference equality the right thing to do here? I'm not sure this is guaranteed.